# <span style="color:#d50283">IT Academy - Data Science amb Python</span>
## S03 T05: Exploració de les dades
**Familiaritza't amb les tècniques d'exploració de les dades mitjantçant la estructura de dades, Dataframe amb la llibreria Pandas.**

***
###  Exercici 1
*Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un pandas Dataframe. Explora les dades que conté, i queda’t únicament amb les columnes que consideris rellevants.*

In [1]:
import pandas as pd
import numpy as np





import numpy; print(f'Numpy version {numpy.__version__}')
import pandas; print(f'Pandas version {pandas.__version__}')



# importamos archivo
df=pd.read_csv("D:/data/DelayedFlights.csv")

Numpy version 1.20.1
Pandas version 1.2.4


**Visión Archivo**

In [2]:
df.head(10)

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   
5           6  2008      1           3          4   1937.0        1830   
6          10  2008      1           3          4    706.0         700   
7          11  2008      1           3          4   1644.0        1510   
8          15  2008      1           3          4   1029.0        1020   
9          16  2008      1           3          4   1452.0        1425   

   ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0   2211.0        2225            WN  ...     4.0     8.0          0   
1   1002.0        1000            WN  ...     5.0    10.0          0   
2    804.0         750            WN  ...     3.0    17.0          0   
3   1959.0        1925            WN  ...     3.0    10.0          0   
4   2121.0        2110            WN  ...     4.0    10.0          0   
5   2037.0        1940            WN  ...     3.0     7.0          0   
6    916.0         915            WN  ...     5.0    19.0          0   
7   1845.0        1725            WN  ...     6.0     8.0          0   
8   1021.0        1010            WN  ...     6.0     9.0          0   
9   1640.0        1625            WN  ...     7.0     8.0          0   

   CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                 N         0           NaN           NaN      NaN   
1                 N         0           NaN           NaN      NaN   
2                 N         0           NaN           NaN      NaN   
3                 N         0           2.0           0.0      0.0   
4                 N         0           NaN           NaN      NaN   
5                 N         0          10.0           0.0      0.0   
6                 N         0           NaN           NaN      NaN   
7                 N         0           8.0           0.0      0.0   
8                 N         0           NaN           NaN      NaN   
9                 N         0           3.0           0.0      0.0   

  SecurityDelay  LateAircraftDelay  
0           NaN                NaN  
1           NaN                NaN  
2           NaN                NaN  
3           0.0               32.0  
4           NaN                NaN  
5           0.0               47.0  
6           NaN                NaN  
7           0.0               72.0  
8           NaN                NaN  
9           0.0               12.0  

[10 rows x 30 columns]

In [3]:
df.tail(10)

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1936748     7009699  2008     12          13          6   1254.0        1221   
1936749     7009701  2008     12          13          6   1528.0        1500   
1936750     7009702  2008     12          13          6   1531.0        1522   
1936751     7009705  2008     12          13          6    921.0         830   
1936752     7009709  2008     12          13          6   1552.0        1520   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
1936748   1420.0        1359            DL  ...     9.0    13.0          0   
1936749   1720.0        1642            DL  ...     4.0    29.0          0   
1936750   1822.0        1823            DL  ...     9.0    14.0          0   
1936751   1112.0        1008            DL  ...     8.0    21.0          0   
1936752   1735.0        1718            DL  ...     9.0     7.0          0   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
1936748                 N         0           0.0           0.0      0.0   
1936749                 N         0          16.0           0.0     10.0   
1936750                 N         0           NaN           NaN      NaN   
1936751                 N         0          51.0           0.0     13.0   
1936752                 N         0           0.0           0.0      0.0   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
1936748           0.0               21.0  
1936749           0.0               12.0  
1936750           NaN                NaN  
1936751           0.0                0.0  
1936752           0.0               17.0  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[10 rows x 30 columns]

In [4]:
# Número de filas y columnas
df.shape

(1936758, 30)

In [5]:
#Type de columnas.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

In [6]:
# Analisis estadistico de las columnas numericas.
df.describe()

Unnamed: 0       Year         Month    DayofMonth     DayOfWeek  \
count  1.936758e+06  1936758.0  1.936758e+06  1.936758e+06  1.936758e+06   
mean   3.341651e+06     2008.0  6.111106e+00  1.575347e+01  3.984827e+00   
std    2.066065e+06        0.0  3.482546e+00  8.776272e+00  1.995966e+00   
min    0.000000e+00     2008.0  1.000000e+00  1.000000e+00  1.000000e+00   
25%    1.517452e+06     2008.0  3.000000e+00  8.000000e+00  2.000000e+00   
50%    3.242558e+06     2008.0  6.000000e+00  1.600000e+01  4.000000e+00   
75%    4.972467e+06     2008.0  9.000000e+00  2.300000e+01  6.000000e+00   
max    7.009727e+06     2008.0  1.200000e+01  3.100000e+01  7.000000e+00   

            DepTime    CRSDepTime       ArrTime    CRSArrTime     FlightNum  \
count  1.936758e+06  1.936758e+06  1.929648e+06  1.936758e+06  1.936758e+06   
mean   1.518534e+03  1.467473e+03  1.610141e+03  1.634225e+03  2.184263e+03   
std    4.504853e+02  4.247668e+02  5.481781e+02  4.646347e+02  1.944702e+03   
min    1.000000e+00  0.000000e+00  1.000000e+00  0.000000e+00  1.000000e+00   
25%    1.203000e+03  1.135000e+03  1.316000e+03  1.325000e+03  6.100000e+02   
50%    1.545000e+03  1.510000e+03  1.715000e+03  1.705000e+03  1.543000e+03   
75%    1.900000e+03  1.815000e+03  2.030000e+03  2.014000e+03  3.422000e+03   
max    2.400000e+03  2.359000e+03  2.400000e+03  2.400000e+03  9.742000e+03   

       ...      Distance        TaxiIn       TaxiOut     Cancelled  \
count  ...  1.936758e+06  1.929648e+06  1.936303e+06  1.936758e+06   
mean   ...  7.656862e+02  6.812975e+00  1.823220e+01  3.268348e-04   
std    ...  5.744797e+02  5.273595e+00  1.433853e+01  1.807562e-02   
min    ...  1.100000e+01  0.000000e+00  0.000000e+00  0.000000e+00   
25%    ...  3.380000e+02  4.000000e+00  1.000000e+01  0.000000e+00   
50%    ...  6.060000e+02  6.000000e+00  1.400000e+01  0.000000e+00   
75%    ...  9.980000e+02  8.000000e+00  2.100000e+01  0.000000e+00   
max    ...  4.962000e+03  2.400000e+02  4.220000e+02  1.000000e+00   

           Diverted  CarrierDelay  WeatherDelay      NASDelay  SecurityDelay  \
count  1.936758e+06  1.247488e+06  1.247488e+06  1.247488e+06   1.247488e+06   
mean   4.003598e-03  1.917940e+01  3.703571e+00  1.502164e+01   9.013714e-02   
std    6.314722e-02  4.354621e+01  2.149290e+01  3.383305e+01   2.022714e+00   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00   
50%    0.000000e+00  2.000000e+00  0.000000e+00  2.000000e+00   0.000000e+00   
75%    0.000000e+00  2.100000e+01  0.000000e+00  1.500000e+01   0.000000e+00   
max    1.000000e+00  2.436000e+03  1.352000e+03  1.357000e+03   3.920000e+02   

       LateAircraftDelay  
count       1.247488e+06  
mean        2.529647e+01  
std         4.205486e+01  
min         0.000000e+00  
25%         0.000000e+00  
50%         8.000000e+00  
75%         3.300000e+01  
max         1.316000e+03  

[8 rows x 25 columns]

In [7]:
#Numero de nulos por columna
df.isnull().sum()

Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

In [8]:
print(df.columns)

Index(['Unnamed: 0', 'Year', 'Month', 'DayofMonth', 'DayOfWeek', 'DepTime',
       'CRSDepTime', 'ArrTime', 'CRSArrTime', 'UniqueCarrier', 'FlightNum',
       'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'AirTime', 'ArrDelay',
       'DepDelay', 'Origin', 'Dest', 'Distance', 'TaxiIn', 'TaxiOut',
       'Cancelled', 'CancellationCode', 'Diverted', 'CarrierDelay',
       'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'],
      dtype='object')


In [9]:
##Introducimos el listados de las columnas
columnas=['Month','DayofMonth','DayOfWeek','UniqueCarrier','FlightNum','AirTime','ArrDelay',\
                          'DepDelay','Origin','Dest','Distance','ActualElapsedTime','Cancelled','Diverted']

In [10]:
df_vuelo= df[columnas].copy()
df_vuelo

Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  AirTime  \
0            1           3          4            WN        335    116.0   
1            1           3          4            WN       3231    113.0   
2            1           3          4            WN        448     76.0   
3            1           3          4            WN       3920     77.0   
4            1           3          4            WN        378     87.0   
...        ...         ...        ...           ...        ...      ...   
1936753     12          13          6            DL       1621    120.0   
1936754     12          13          6            DL       1631     78.0   
1936755     12          13          6            DL       1631    122.0   
1936756     12          13          6            DL       1639     89.0   
1936757     12          13          6            DL       1641    104.0   

         ArrDelay  DepDelay Origin Dest  Distance  ActualElapsedTime  \
0           -14.0       8.0    IAD  TPA       810              128.0   
1             2.0      19.0    IAD  TPA       810              128.0   
2            14.0       8.0    IND  BWI       515               96.0   
3            34.0      34.0    IND  BWI       515               90.0   
4            11.0      25.0    IND  JAX       688              101.0   
...           ...       ...    ...  ...       ...                ...   
1936753      25.0      30.0    MSP  ATL       906              147.0   
1936754      75.0      57.0    RIC  ATL       481              127.0   
1936755      99.0      80.0    ATL  IAH       689              162.0   
1936756       9.0      11.0    IAD  ATL       533              115.0   
1936757      -5.0       7.0    SAT  ATL       874              123.0   

         Cancelled  Diverted  
0                0         0  
1                0         0  
2                0         0  
3                0         0  
4                0         0  
...            ...       ...  
1936753          0         0  
1936754          0         0  
1936755          0         0  
1936756          0         0  
1936757          0         0  

[1936758 rows x 14 columns]

***
### Exercici 2

Fes un informe complet del data set:.

* Resumeix estadísticament les columnes d’interès
* Troba quantes dades faltants hi ha per columna
* Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...)
* Taula de les aerolínies amb més endarreriments acumulats
* Quins són els vols més llargs? I els més endarrerits?
* Etc

In [11]:
df_vuelo.describe()

Month    DayofMonth     DayOfWeek     FlightNum       AirTime  \
count  1.936758e+06  1.936758e+06  1.936758e+06  1.936758e+06  1.928371e+06   
mean   6.111106e+00  1.575347e+01  3.984827e+00  2.184263e+03  1.082771e+02   
std    3.482546e+00  8.776272e+00  1.995966e+00  1.944702e+03  6.864261e+01   
min    1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    3.000000e+00  8.000000e+00  2.000000e+00  6.100000e+02  5.800000e+01   
50%    6.000000e+00  1.600000e+01  4.000000e+00  1.543000e+03  9.000000e+01   
75%    9.000000e+00  2.300000e+01  6.000000e+00  3.422000e+03  1.370000e+02   
max    1.200000e+01  3.100000e+01  7.000000e+00  9.742000e+03  1.091000e+03   

           ArrDelay      DepDelay      Distance  ActualElapsedTime  \
count  1.928371e+06  1.936758e+06  1.936758e+06       1.928371e+06   
mean   4.219988e+01  4.318518e+01  7.656862e+02       1.333059e+02   
std    5.678472e+01  5.340250e+01  5.744797e+02       7.206007e+01   
min   -1.090000e+02  6.000000e+00  1.100000e+01       1.400000e+01   
25%    9.000000e+00  1.200000e+01  3.380000e+02       8.000000e+01   
50%    2.400000e+01  2.400000e+01  6.060000e+02       1.160000e+02   
75%    5.600000e+01  5.300000e+01  9.980000e+02       1.650000e+02   
max    2.461000e+03  2.467000e+03  4.962000e+03       1.114000e+03   

          Cancelled      Diverted  
count  1.936758e+06  1.936758e+06  
mean   3.268348e-04  4.003598e-03  
std    1.807562e-02  6.314722e-02  
min    0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  
75%    0.000000e+00  0.000000e+00  
max    1.000000e+00  1.000000e+00

In [12]:
df_vuelo.isnull().sum()

Month                   0
DayofMonth              0
DayOfWeek               0
UniqueCarrier           0
FlightNum               0
AirTime              8387
ArrDelay             8387
DepDelay                0
Origin                  0
Dest                    0
Distance                0
ActualElapsedTime    8387
Cancelled               0
Diverted                0
dtype: int64

*Los nulos corresponden a los vuelos cancelados y desviados*

In [13]:
df_vuelo['Cancelled'].sum()+df_vuelo['Diverted'].sum()

8387

In [14]:
df_vuelo['UniqueCarrier'].describe()

count     1936758
unique         20
top            WN
freq       377602
Name: UniqueCarrier, dtype: object

In [15]:
cias=df_vuelo['UniqueCarrier'].describe()

print(f'En 2018 se han realizado {cias[0]} vuelos, lo han realizado {cias[1]} compañias,\n\
la compañia que más vuelos a realizado es {cias[2]} que ha realizado {cias[3]} vuelos')

En 2018 se han realizado 1936758 vuelos, lo han realizado 20 compañias,
la compañia que más vuelos a realizado es WN que ha realizado 377602 vuelos


In [16]:
df['UniqueCarrier'].value_counts(sort=True)

WN    377602
AA    191865
MQ    141920
UA    141426
OO    132433
DL    114238
XE    103663
CO    100195
US     98425
EV     81877
NW     79108
FL     71284
YV     67063
B6     55315
OH     52657
9E     51885
AS     39293
F9     28269
HA      7490
AQ       750
Name: UniqueCarrier, dtype: int64

In [17]:
pd.set_option('display.float_format', '{:.2%}'.format) ## cambiamos el formato de los 
df['UniqueCarrier'].value_counts(sort=True,normalize=True)


WN   19.50%
AA    9.91%
MQ    7.33%
UA    7.30%
OO    6.84%
DL    5.90%
XE    5.35%
CO    5.17%
US    5.08%
EV    4.23%
NW    4.08%
FL    3.68%
YV    3.46%
B6    2.86%
OH    2.72%
9E    2.68%
AS    2.03%
F9    1.46%
HA    0.39%
AQ    0.04%
Name: UniqueCarrier, dtype: float64

In [18]:
pd.reset_option('display.float_format')
Origen=df_vuelo['Origin'].describe()

print(f'En 2018 se han realizado {Origen[0]} vuelos, desde   {Origen[1]} origenes diferentes,\n\
el origen desde donde más se ha despegado es {Origen[2]} desde don han salido  {Origen[3]} vuelos.')

En 2018 se han realizado 1936758 vuelos, desde   303 origenes diferentes,
el origen desde donde más se ha despegado es ATL desde don han salido  131613 vuelos.


In [19]:
Dest=df_vuelo['Dest'].describe()

print(f'En 2018 se han realizado {Dest[0]} vuelos, con destino a  {Dest[1]} destinos diferentes ,\n\
el destino donde más se ha aterrizado  es {Dest[2]} donde han llegado  {Dest[3]} vuelos.')

En 2018 se han realizado 1936758 vuelos, con destino a  304 destinos diferentes ,
el destino donde más se ha aterrizado  es ORD donde han llegado  108984 vuelos.


In [20]:
df_vuelo['FlightNum'].describe()

count    1.936758e+06
mean     2.184263e+03
std      1.944702e+03
min      1.000000e+00
25%      6.100000e+02
50%      1.543000e+03
75%      3.422000e+03
max      9.742000e+03
Name: FlightNum, dtype: float64

In [21]:
vuelo_cont=df_vuelo['FlightNum'].value_counts(sort=True,ascending=False)
vuelo_cont

16      1586
50      1550
40      1507
15      1471
44      1447
        ... 
6947       1
6870       1
6191       1
6249       1
7702       1
Name: FlightNum, Length: 7499, dtype: int64

In [22]:
vuel=len(df_vuelo['FlightNum'].unique())
print(f'Los vuelos se han organizado en {vuel} números diferentes el vuelo con mas frecuencia es el {vuelo_cont.index[0]}')

Los vuelos se han organizado en 7499 números diferentes el vuelo con mas frecuencia es el 16


### Añadidmos columnas
**Retraso:**
* 0 no hay retraso
* 1 Si hay retraso.



In [23]:
## definimos el retraso como mayor a 30 minutos. También los Desviados y cancelados aparecen como retraso.
def retraso_bin(x):
    
    '''Se define como retraso encima de 30 min más 
    los vuelos desviados y candelados'''
    
    if x>=30.0:
        retraso=1
    
    elif x<30:
        retraso=0
    else:
        retraso=1
        
  
    return retraso

df_vuelo["Retraso"]=df['ArrDelay'].apply(retraso_bin)

df_vuelo.head()

Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  AirTime  ArrDelay  \
0      1           3          4            WN        335    116.0     -14.0   
1      1           3          4            WN       3231    113.0       2.0   
2      1           3          4            WN        448     76.0      14.0   
3      1           3          4            WN       3920     77.0      34.0   
4      1           3          4            WN        378     87.0      11.0   

   DepDelay Origin Dest  Distance  ActualElapsedTime  Cancelled  Diverted  \
0       8.0    IAD  TPA       810              128.0          0         0   
1      19.0    IAD  TPA       810              128.0          0         0   
2       8.0    IND  BWI       515               96.0          0         0   
3      34.0    IND  BWI       515               90.0          0         0   
4      25.0    IND  JAX       688              101.0          0         0   

   Retraso  
0        0  
1        0  
2        0  
3        1  
4        0

In [24]:
num_retraso=df_vuelo['Retraso'].value_counts(normalize=True)
num_retraso.head()

0    0.556263
1    0.443737
Name: Retraso, dtype: float64

In [25]:
 ## cambiamos el formato de lo
print("En el 2018 un {:2.2%}  de los vuelos han tenido retraso".format(num_retraso[0]))


En el 2018 un 55.63%  de los vuelos han tenido retraso


In [26]:
df_vuelo['Velocidad']=df_vuelo['Distance']/(df_vuelo['AirTime']/60)
df_vuelo['Velocidad'].round(2)

0          418.97
1          430.09
2          406.58
3          401.30
4          474.48
            ...  
1936753    453.00
1936754    370.00
1936755    338.85
1936756    359.33
1936757    504.23
Name: Velocidad, Length: 1936758, dtype: float64

In [27]:
## Eliminamo los valores infinitos. Lo sustitumos por nan

df_vuelo.replace([np.inf, -np.inf], np.nan, inplace=True)

In [28]:
df_vuelo.head()

Month  DayofMonth  DayOfWeek UniqueCarrier  FlightNum  AirTime  ArrDelay  \
0      1           3          4            WN        335    116.0     -14.0   
1      1           3          4            WN       3231    113.0       2.0   
2      1           3          4            WN        448     76.0      14.0   
3      1           3          4            WN       3920     77.0      34.0   
4      1           3          4            WN        378     87.0      11.0   

   DepDelay Origin Dest  Distance  ActualElapsedTime  Cancelled  Diverted  \
0       8.0    IAD  TPA       810              128.0          0         0   
1      19.0    IAD  TPA       810              128.0          0         0   
2       8.0    IND  BWI       515               96.0          0         0   
3      34.0    IND  BWI       515               90.0          0         0   
4      25.0    IND  JAX       688              101.0          0         0   

   Retraso   Velocidad  
0        0  418.965517  
1        0  430.088496  
2        0  406.578947  
3        1  401.298701  
4        0  474.482759

In [29]:
df_vuelo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 16 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Month              int64  
 1   DayofMonth         int64  
 2   DayOfWeek          int64  
 3   UniqueCarrier      object 
 4   FlightNum          int64  
 5   AirTime            float64
 6   ArrDelay           float64
 7   DepDelay           float64
 8   Origin             object 
 9   Dest               object 
 10  Distance           int64  
 11  ActualElapsedTime  float64
 12  Cancelled          int64  
 13  Diverted           int64  
 14  Retraso            int64  
 15  Velocidad          float64
dtypes: float64(5), int64(8), object(3)
memory usage: 236.4+ MB


**Las compañias más rapidas:**

In [30]:
velocidad_cia=df_vuelo.groupby('UniqueCarrier')['Velocidad'].mean().round(2).to_frame()
velocidad_cia.sort_values(by='Velocidad',ascending=False)

Velocidad
UniqueCarrier           
F9                437.61
AA                435.58
CO                432.75
UA                432.00
AS                418.43
DL                417.48
B6                417.03
FL                410.00
US                409.80
NW                402.79
WN                396.55
XE                382.02
HA                373.05
EV                370.88
AQ                369.76
9E                369.29
OH                358.36
MQ                353.88
YV                347.98
OO                342.88

**Los vuelos más rapidos**

In [31]:

velocidad_vuelos=df_vuelo.groupby('FlightNum')['Velocidad'].mean().round(2).to_frame()
velocidad_vuelos.sort_values(by='Velocidad',ascending=False).head(10)


Velocidad
FlightNum           
4200          729.04
5100          579.69
7441          532.80
7449          528.89
7453          525.72
254           522.91
7700          522.36
7367          511.74
5996          509.59
6116          508.50

### Analisis de los retrasos
#### Retrasos por:
* Dia Total

In [32]:
df_vuelo.groupby('DayOfWeek').agg({'Retraso':sum,'FlightNum':len}).sort_values(by='Retraso',ascending=False)

Retraso  FlightNum
DayOfWeek                    
5           149289     323259
7           129458     286111
1           128432     290933
4           126498     289451
2           118751     260943
3           114320     262805
6            92664     223256

* Dia %

In [33]:
dia=df_vuelo.groupby(['DayOfWeek','Retraso']).agg(len)

dia['Porcentaje']=df_vuelo.groupby('DayOfWeek')["Retraso"].value_counts(normalize=True)

dia.reset_index(level=1, inplace=True)
#prueba['Cuenta_Retraso']=prueba.index
dia_sim=dia[dia['Retraso']==1].sort_values(by='Porcentaje',ascending=False)
pd.set_option('display.float_format', '{:.2%}'.format) ## cambiamos el formato de los
dia_sim[['Retraso','Porcentaje']]

Retraso  Porcentaje
DayOfWeek                     
5                1      46.18%
2                1      45.51%
7                1      45.25%
1                1      44.14%
4                1      43.70%
3                1      43.50%
6                1      41.51%

In [34]:
pd.reset_option('display.float_format')

* Mes Total:

In [35]:
df_vuelo.groupby('Month').agg({'Retraso':sum,'FlightNum':len}).sort_values(by='Retraso',ascending=False)

Retraso  FlightNum
Month                    
12      103557     203385
6        97486     200914
2        91064     189534
3        89093     200842
7        84449     182945
1        83192     183527
8        69272     162648
4        64651     155264
5        61891     153072
11       43154     105563
9        36247      95539
10       35356     103525

* Mes %

In [36]:
month=df_vuelo.groupby(['Month','Retraso']).agg(len)

month['Porcentaje']=df_vuelo.groupby('Month')["Retraso"].value_counts(normalize=True)

month.reset_index(level=1, inplace=True)
#prueba['Cuenta_Retraso']=prueba.index
month_sim=month[month['Retraso']==1].sort_values(by='Porcentaje',ascending=False)
pd.set_option('display.float_format', '{:.2%}'.format) ## cambiamos el formato de los
month_sim[['Porcentaje']]

Porcentaje
Month            
12         50.92%
6          48.52%
2          48.05%
7          46.16%
1          45.33%
3          44.36%
8          42.59%
4          41.64%
11         40.88%
5          40.43%
9          37.94%
10         34.15%

* Compañía Total

In [37]:
df_vuelo.groupby('UniqueCarrier').agg({'Retraso':sum,'FlightNum':len}).sort_values(by='Retraso',ascending=False)

Retraso  FlightNum
UniqueCarrier                    
WN              125555     377602
AA               95433     191865
UA               70127     141426
MQ               68872     141920
OO               62504     132433
XE               52297     103663
DL               47627     114238
CO               41928     100195
EV               39842      81877
US               39826      98425
YV               36770      67063
NW               36710      79108
FL               31776      71284
B6               29668      55315
OH               28398      52657
9E               25509      51885
AS               15665      39293
F9                8433      28269
HA                2293       7490
AQ                 179        750

* Compañia %

In [38]:
cia=df_vuelo.groupby(['UniqueCarrier','Retraso']).agg(len)

cia['Porcentaje']=df_vuelo.groupby('UniqueCarrier')["Retraso"].value_counts(normalize=True)

cia.reset_index(level=1, inplace=True)
#prueba['Cuenta_Retraso']=prueba.index
cia_sim=cia[cia['Retraso']==1].sort_values(by='Porcentaje',ascending=False)
pd.set_option('display.float_format', '{:.2%}'.format) ## cambiamos el formato de los
cia_sim[['Porcentaje']]

Porcentaje
UniqueCarrier            
YV                 54.83%
OH                 53.93%
B6                 53.63%
XE                 50.45%
AA                 49.74%
UA                 49.59%
9E                 49.16%
EV                 48.66%
MQ                 48.53%
OO                 47.20%
NW                 46.40%
FL                 44.58%
CO                 41.85%
DL                 41.69%
US                 40.46%
AS                 39.87%
WN                 33.25%
HA                 30.61%
F9                 29.83%
AQ                 23.87%

In [39]:
cia_total=df_vuelo.groupby('UniqueCarrier').agg({'Retraso':sum,'FlightNum':len})
cia=df_vuelo.groupby(['UniqueCarrier','Retraso']).agg(len)

cia['Porcentaje']=df_vuelo.groupby('UniqueCarrier')["Retraso"].value_counts(normalize=True)

cia.reset_index(level=1, inplace=True)
#prueba['Cuenta_Retraso']=prueba.index
cia_sim=cia[cia['Retraso']==1].sort_values(by='Porcentaje',ascending=False)
pd.set_option('display.float_format', '{:.2%}'.format) ## cambiamos el formato de los
cia_total['Porcentaje']=cia_sim['Porcentaje']
cia_total.rename(columns={'Retraso':'N Vuelos Retrasados','FlightNum':'Total Vuelos','Porcentaje':'Porcentaje_Retrasos'}, inplace=True)
cia_total.sort_values(by='Porcentaje_Retrasos',ascending=False)

N Vuelos Retrasados  Total Vuelos  Porcentaje_Retrasos
UniqueCarrier                                                        
YV                           36770         67063               54.83%
OH                           28398         52657               53.93%
B6                           29668         55315               53.63%
XE                           52297        103663               50.45%
AA                           95433        191865               49.74%
UA                           70127        141426               49.59%
9E                           25509         51885               49.16%
EV                           39842         81877               48.66%
MQ                           68872        141920               48.53%
OO                           62504        132433               47.20%
NW                           36710         79108               46.40%
FL                           31776         71284               44.58%
CO                           41928        100195               41.85%
DL                           47627        114238               41.69%
US                           39826         98425               40.46%
AS                           15665         39293               39.87%
WN                          125555        377602               33.25%
HA                            2293          7490               30.61%
F9                            8433         28269               29.83%
AQ                             179           750               23.87%

* Num vuelo:

In [40]:
df_vuelo.groupby('FlightNum').agg({'Retraso':sum,'FlightNum':len}).sort_values(by='Retraso',ascending=False).head(10)

Retraso  FlightNum
FlightNum                    
50             726       1550
321            689       1355
75             665       1354
378            657       1238
36             624       1260
52             617       1410
16             616       1586
55             614       1278
44             603       1447
40             603       1507

* Num vuelo %:
    

In [41]:
vuel_num=df_vuelo.groupby(['FlightNum','Retraso']).agg(len)

vuel_num['Porcentaje']=df_vuelo.groupby('FlightNum')["Retraso"].value_counts(normalize=True)

vuel_num.reset_index(level=1, inplace=True)
#prueba['Cuenta_Retraso']=prueba.index
vuel_num_sim=vuel_num[vuel_num['Retraso']==1].sort_values(by='Porcentaje',ascending=False)
pd.set_option('display.float_format', '{:.2%}'.format) ## cambiamos el formato de los
vuel_num_sim[['Porcentaje']]

Porcentaje
FlightNum            
9742          100.00%
7452          100.00%
7376          100.00%
6899          100.00%
7396          100.00%
...               ...
1385           11.76%
3289           11.43%
3298            9.52%
7715            7.69%
3288            7.14%

[7461 rows x 1 columns]

#### Los diez vuelos más largos

In [42]:
pd.reset_option('display.float_format')

df_vuelo.groupby('FlightNum').agg({'Distance':'mean','FlightNum':len}).sort_values(by='Distance',ascending=False).head(10)

Distance  FlightNum
FlightNum                        
1          2195.572093        860
2          2088.130964        985
73         2033.562216       1101
15         2032.486744       1471
181        1861.228731        717
83         1842.425926        594
1829       1837.040284        422
1273       1827.247126        696
14         1774.276644        882
17         1722.519651       1145

***

### Exercici 3

*Exporta el data set net i amb les noves columnes a Excel.*

*This sheet is too large! Your sheet size is: 1936758, 16 Max sheet size is: 1048576, 16384*

El archivo es muy largo. Lo separaremos en 4 archivos. 1 por cuatrimestre


In [43]:
flight_1q=df_vuelo[df_vuelo['Month']<5]
flight_2q=df_vuelo[(df_vuelo['Month']>4)& (df_vuelo['Month']<8)]
flight_3q=df_vuelo[(df_vuelo['Month']>7)& (df_vuelo['Month']<10)]
flight_4q=df_vuelo[(df_vuelo['Month']>9)]



In [44]:
flight_1q.to_excel('flight_14q.xlsx',sheet_name='1Q2018')
flight_2q.to_excel('flight_2q.xlsx',sheet_name='2Q2018')
flight_3q.to_excel('flight_3q.xlsx',sheet_name='3Q2018')
flight_4q.to_excel('flight_4q.xlsx',sheet_name='4Q2018')

In [45]:
cia_total.to_excel('Retrasos_Compania.xlsx', sheet_name="Retrasos_Cia")